packages/datadog-plugin-kafkajs/test/index.spec.js

```diff
@@ -0,0 +1,139 @@
+  describe('kafkajs', function () {
+    this.timeout(TIMEOUT)
+    afterEach(() => {
+      agent.close()
+      agent.wipe()
+    })
+    withVersions(plugin, 'kafkajs', (version) => {
+      const testTopic = 'topic-test'
+      let kafka
+      let tracer
+      describe('without configuration', () => {
+        beforeEach(async () => {
+          tracer = require('../../dd-trace')
+          agent.load('kafkajs')
+          const {
+            Kafka
+          } = require(`../../../versions/kafkajs@${version}`).get()
+          kafka = new Kafka({
+            clientId: 'kafkajs-test',
+            brokers: [`localhost:9092`]
+          })
+        })
+        describe('producer', () => {
+          const messages = [{ key: 'key1', value: 'test2' }, { key: 'key2', value: 'test2' }]
+
+          it('should be instrumented', async () => {
+            const producer = kafka.producer()
+            try {
+              const expectedSpanPromise = expectSpanWithDefaults({
+                name: 'kafka.produce',
+                service: 'test-kafka',
+                meta: {
+                  'span.kind': 'producer',
+                  'component': 'kafka'
+                },
+                metrics: {
+                  'kafka.batch.size': messages.length
+                },
+                resource: testTopic,
+                error: 0,
+                type: 'queue'
+
+              })
+
+              await producer.connect()
+              await producer.send({
+                topic: testTopic,
+                messages
+              })
+              // Useful to debug trace
+              // agent.use(traces => console.log(traces[0]))
+
+              return expectedSpanPromise
+            } catch (error) {
+              // console.log(error)
+            }
+          })
+          it('should propagate context', async () => {
+            const producer = kafka.producer()
+            const firstSpan = tracer.scope().active()
+            await producer.connect()
+            await producer.send({
+              topic: testTopic,
+              messages: [{ key: 'key1', value: 'test' }]
+            })
+
+            return expect(tracer.scope().active()).to.equal(firstSpan)
+          })
+        })
+        describe('consumer', () => {
+          it('should be instrumented', async () => {
+            // We are changing the groupId in every test so the consumed offset is always reset
+            const consumer = kafka.consumer({ groupId: `test-group-${version}-instrument` })
+            try {
+              const expectedSpanPromise = expectSpanWithDefaults({
+                name: 'kafka.consume',
+                service: 'test-kafka',
+                meta: {
+                  'span.kind': 'consumer',
+                  'component': 'kafka'
+                },
+                resource: testTopic,
+                error: 0,
+                type: 'queue'
+              })
+
+              await consumer.connect()
```

Comment:
This and the following line should be moved to a `beforeEach` with some unique identifier between runs. This will allow cleaning up in an `afterEach` and avoid possible side-effects between runs.

Reply:
Each test has its own different consumer that is created inside that test. This allows the different consumers to have different offsets, preventing them of influencing each other

Reply:
This is still possible even with a `beforeEach`. The problem with the current approach is that an error will cause the consumer to never disconnect since it doesn't happen in an `afterEach` which would guarantee that the code would run regardless of any errors in each test.

Reply:
Oh! I see what you mean. I'll get to it tomorrow!